text/3481-special-cased-unsized-sums.md

```diff
@@ -0,0 +1,106 @@
+- Feature Name: special_cased_unsized_sums
+- Start Date: 2023-08-28
+- RFC PR: [rust-lang/rfcs#3481](https://github.com/rust-lang/rfcs/pull/3481)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Currently `Iterator::sum` does a relatively slow iterative sum using 
+`Iterator::fold` internally. For sequences of positive integers, the sum of the sequence 
+has been proven down to a simple formula that computes much faster than an iterative sum.
+This RFC suggests a special-case implementation of `Iterator::sum` for `Range<u*>` and 
+`RangeInclusive<u*>` (where `u*` is `u8`, `u16`, `u32`, `u64`, `u128`, and `usize`). 
+
+The base formula to be used is available on Wikipedia [here](https://en.wikipedia.org/wiki/1_%2B_2_%2B_3_%2B_4_%2B_%E2%8B%AF).
+
+# Motivation
+[motivation]: #motivation
+
+This should signifigantly improve the performance of any code that uses some variant of 
+`(0..n).sum()`. This may be somewhat uncommon, but is still worth the performance increase for 
+the small number of cases. 
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+For any consecutive sequence of positive integers from 1 to n inclusive (`1..=n`), the sum will 
+equal `n*(n+1)/2`. This also works for ranges from 0 to n (inclusive; `0..=n`) because 
+`sum(0..=n) == 0 + sum(1..=n)`. We can generalize this to work for any range by simply 
```

Comment:
Just a question about the scope of this change: why aren't we specialising for signed integer types as well?

Here's an example of guide text for signed ranges:
```
We can generalize this to work for a signed range by taking the sum of the positive range, then subtracting the sum of the absolute value of the negative range. Then the sums of both ranges are calculated using the positive sum function.

For example, to do `sum(-4..=10)` we would do 
`sum(1..=10) - sum(1..=4)`: `10*11/2 - 4*5/2 == 55-10 == 45 == -4-3-2-1+0+1+2+3+4+5+6+7+8+9+10 == sum(-4..=10)`.
```

Handling ranges that include the minimum signed value requires converting the absolute value to the unsigned type of the same width. As an alternative, these ranges can be special-cased, because there are only two signed sums starting at the minimum value where the naive sum doesn't underflow: `i128::MIN` itself, and `i128::MIN + (i128::MIN - 1)`.

Reply:
This formula also preserves the same overflow behaviour, as long as the positive and negative sums are cast back to the original type before being added together. (And the negative sun is negated before the cast.)